Firestore/Swift/Source/Codable/DocumentReference+ReadDecodable.swift

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFirestore
+
+extension DocumentReference {
+  /// Encodes an instance of `Encodable` and overwrites the encoded data
+  /// to the document referred by this `DocumentReference`. If no document exists,
+  /// it is created. If a document already exists, it is overwritten.
+  ///
+  /// See `Firestore.Encoder` for more details about the encoding process.
+  ///
+  /// - Parameters:
+  ///   - value: An instance of `Encodable` to be encoded to a document.
+  ///   - encoder: An encoder instance to use to run the encoding.
+  ///   - completion: A block to execute once the document has been successfully
+  ///                 written to the server. This block will not be called while
+  ///                 the client is offline, though local changes will be visible
+  ///                 immediately.
+
+  public func getDocument<T: Decodable>(as type: T.Type,
+                                        with serverTimestampBehavior: ServerTimestampBehavior =
+                                          .none,
+                                        decoder: Firestore.Decoder? = nil,
+                                        completion: @escaping (Result<T, Error>) -> Void) {
+    getDocument { snapshot, error in
+      guard let snapshot = snapshot else {
+        completion(.failure(error ?? FirestoreDecodingError.internal))
```

Comment:
@schmidt-sebastian if we decide to introduce a new error constant to indicate internal errors, we should make sure to update other similar locations as well.

Reply:
Another solution could be to fatalError - in case the underlying api guarantees that either error or value is present.

Reply:
I guess something similar to a `fatalError()` or other means of trapping must be present in the auto-synthesised async interfaces from Objective-C methods with completion handlers with an optional value and an optional error. 
So on that account it would seem 'safe' just to `fatalError()` here too. What do you think?

Reply:
Please correct me if I'm wrong but this introduces a fork in the functionality with the async/await version, right? The async/await version has the same issue where it expects the API to honour the contract or else it'll have unexpected behaviour. 

Firestore guarantees that only one of `snapshot` or `error` will be instantiated and we should expect that - I think it's fine to force unwrap the `error` here to avoid introducing a new public API for something that should never happen. If this ever crashes, I'm happy to take the blame 😛 

.... I love that this isn't something we have to think about when consuming Swift APIs 😄

Reply:
Updated note after chatting with @paulb777: we can test the async/await behaviour by hardcoding `nil/nil` and see what happens and try to have a reasonable behaviour between the two.

Reply:
Thanks for adding an inline comment documenting the behaviour of the code, @mortenbekditlevsen.

From my perspective, this LGTM.
@schmidt-sebastian, what's your perspective?